### PR TITLE
RSDK-2703 - Document existing wrappers

### PR DESCRIPTION
**Pull request by stuqdog**
#### Major Changes
- Provides more comprehensive, doxygen-formatted documentation for `robot`, `camera`, `generic`, `rpc/server`, and `registry`

#### Minor changes
- Make use of `should_close_channel` in robot client
- make `threads` `private` in robot client
-

### CODE REVIEW

File: src/robot/client.hpp [hunk 1]
```diff
@@ -81,29 +95,53 @@ class RobotClient {
...
                                      std::unordered_map<std::string, std::shared_ptr<ProtoType>>,
                                      ResourceNameHasher,
                                      ResourceNameEqual> extra);
+
+    /// @brief Cancel a specified operation on the robot.
+    /// @param id The ID of the operation to cancel.
     void cancel_operation(std::string id);
 
    private:
+    std::vector<std::shared_ptr<std::thread>> threads_;
```
Comment (stuqdog) [inline src/robot/client.hpp:144]: (driveby) no one outside of robot client was actually using `threads` so it was trivial to make it private.

File: src/rpc/server.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ void Server::register_service(grpc::Service* service) {
 
 void Server::start() {
     if (server_) {
-        throw "Attempted to start server that was already running";
+        throw std::runtime_error("Attempted to start server that was already running");
```
Comment (stuqdog) [inline src/rpc/server.cpp:22]: (driveby) adding documentation, wanted to throw an actual error rather than a string. easy enough to fix!

**Review by cheukt — APPROVED** [review]
LGTM - not sure if this pr does it, but is there a way to group all related classes together in the class list page (e.g. Camera client, server, and Camera in one dropdown)

**Comment by stuqdog** [comment]
> LGTM - not sure if this pr does it, but is there a way to group all related classes together in the class list page (e.g. Camera client, server, and Camera in one dropdown)

Good question! This pr does not do that, but I'll do some research into it.

**Comment by acmorrow** [comment]
> > LGTM - not sure if this pr does it, but is there a way to group all related classes together in the class list page (e.g. Camera client, server, and Camera in one dropdown)
> 
> Good question! This pr does not do that, but I'll do some research into it.

One thing that comes to mind is using namespaces or nesting more aggressively. Things overall still need to get wrapped in a top-level namespace, and usually I recommend not using complex namespacing schemes, but in principle you could have a `camera` namespace under whatever is the top-level. What I don't love about that is you end up with `camera::Camera`, which feels chatty. But I do maybe like `camera::Server` and `camera::Client` over `CameraServer` and `CameraClient`?

Another idea might be to nest the `Client` and `Server` under `Camera`:

```
class Camera {
public:
    class Server;
    class Client;
};

class Camera::Server : public ... {
};

class Camera::Client : public Camera {
};
```

The definitions of `Camera::Server` and `Camera::Client` don't even need to be in the same header file once you have forward declared them as nested. On the other hand, doing it this way makes it impossible to forward declare the camera server / client since classes are closed and namespaces are open. So I think namespaces are probably a better choice.

Finally, doxygen itself has some grouping features: https://www.doxygen.nl/manual/grouping.html

**Comment by stuqdog** [comment]
> > > LGTM - not sure if this pr does it, but is there a way to group all related classes together in the class list page (e.g. Camera client, server, and Camera in one dropdown)
> > 
> > 
> > Good question! This pr does not do that, but I'll do some research into it.
> 
> One thing that comes to mind is using namespaces or nesting more aggressively. Things overall still need to get wrapped in a top-level namespace, and usually I recommend not using complex namespacing schemes, but in principle you could have a `camera` namespace under whatever is the top-level. What I don't love about that is you end up with `camera::Camera`, which feels chatty. But I do maybe like `camera::Server` and `camera::Client` over `CameraServer` and `CameraClient`?
> 
> Another idea might be to nest the `Client` and `Server` under `Camera`:
> 
> ```
> class Camera {
> public:
>     class Server;
>     class Client;
> };
> 
> class Camera::Server : public ... {
> };
> 
> class Camera::Client : public Camera {
> };
> ```
> 
> The definitions of `Camera::Server` and `Camera::Client` don't even need to be in the same header file once you have forward declared them as nested. On the other hand, doing it this way makes it impossible to forward declare the camera server / client since classes are closed and namespaces are open. So I think namespaces are probably a better choice.
> 
> Finally, doxygen itself has some grouping features: https://www.doxygen.nl/manual/grouping.html

Thanks for the guidance! Adding `namespace`ing is another quick PR I hope to get out by early next week. Not that I necessarily want to go deep into the "nested namespace" rabbithole, but does Doxygen respect/group `namespace`s by default?

**Comment by stuqdog** [comment]
@cheukt just added `group`s for `camera`, `robot`, and `generic`. The class list page is unchanged, but it'll create a new tab (`modules`) which shows that grouping.

**Review by zaporter-work — APPROVED** [review]
LGTM
